### PR TITLE
Use multiple dictionaries for line/word segmenter.

### LINKED ISSUE #1033: Add dictionary based segmenter
For https://github.com/unicode-org/icu4x/issues/813 as CJ

### PR DESCRIPTION
**Pull request by makotokato**
Fixes #1033

### CODE REVIEW

File: experimental/segmenter/src/line.rs [hunk 1]
```diff
@@ -99,7 +99,10 @@ impl Default for LineBreakOptions {
 pub struct LineBreakSegmenter {
     options: LineBreakOptions,
     payload: DataPayload<LineBreakDataV1Marker>,
-    dictionary_payload: DataPayload<UCharDictionaryBreakDataV1Marker>,
+    dictionary_km_payload: Option<DataPayload<UCharDictionaryBreakDataV1Marker>>,
```
Comment (sffc) [inline experimental/segmenter/src/line.rs:102]: Request: Please make LineBreakSegmenter work with either dictionary or LSTM

File: experimental/segmenter/src/line.rs [hunk 2]
```diff
@@ -99,7 +99,10 @@ impl Default for LineBreakOptions {
...
+    dictionary_lo_payload: Option<DataPayload<UCharDictionaryBreakDataV1Marker>>,
```
Comment (sffc) [inline experimental/segmenter/src/line.rs:103]: Question: No CJK dict here?

File: provider/datagen/src/transform/segmenter/mod.rs [hunk 3]
```diff
@@ -337,11 +340,23 @@ impl SegmenterRuleProvider {
                             for c in 0..0x20000 {
                                 if lb.get_u32(c) == LineBreak::ComplexContext {
                                     properties_map[c as usize] = property_index
+                                } else if let Some(c) = char::from_u32(c) {
+                                    match script.get(c) {
+                                        Script::Han | Script::Katakana | Script::Hiragana => {
+                                            properties_map[c as usize] = property_index;
```
Comment (sffc) [inline provider/datagen/src/transform/segmenter/mod.rs:346]: Nit: This change looks like it can be decoupled from the rest of the PR. Please move it to its own PR.

File: experimental/segmenter/src/line.rs (hunk 2 above)
Comment (makotokato) [reply experimental/segmenter/src/line.rs:103]: CJ dictionary is unnecessary for line segmenter since UAX#14 has CJ rules.

File: experimental/segmenter/src/line.rs (hunk 1 above)
Comment (makotokato) [reply experimental/segmenter/src/line.rs:102]: I will add `[cfg(feature = "lstm")]` or not.

File: provider/datagen/src/transform/segmenter/mod.rs (hunk 3 above)
Comment (makotokato) [reply provider/datagen/src/transform/segmenter/mod.rs:346]: OK, I will move mixed text (letter and Chinese) to other PR.

File: experimental/segmenter/src/complex.rs [hunk 4]
```diff
@@ -13,9 +13,26 @@ use crate::DictionarySegmenter;
 #[cfg(feature = "lstm")]
 use crate::lstm::{get_best_lstm_model, LstmSegmenter};
 
+fn get_best_dictionary_paylod(
```
Comment (robertbastian) [inline experimental/segmenter/src/complex.rs:16]: Typo

File: experimental/segmenter/src/complex.rs [hunk 5]
```diff
@@ -13,9 +13,26 @@ use crate::DictionarySegmenter;
...
+    input: u32,
+    payload: [Option<&DataPayload<UCharDictionaryBreakDataV1Marker>>; 5],
```
Comment (robertbastian) [inline experimental/segmenter/src/complex.rs:18]: Suggestion: `[Option<&DataPayload<UCharDictionaryBreakDataV1Marker>>; 5]` isn't a nice type, and it's very easy to populate the fields in an incorrect order. Use a struct instead:

```rust
struct Dictionaries {
  khmer: Option<&DataPayload<UCharDictionaryBreakDataV1Marker>>,
  lao: Option<&DataPayload<UCharDictionaryBreakDataV1Marker>>,
  burmese: Option<&DataPayload<UCharDictionaryBreakDataV1Marker>>,
  thai: Option<&DataPayload<UCharDictionaryBreakDataV1Marker>>,
  cj: Option<&DataPayload<UCharDictionaryBreakDataV1Marker>>,
}

impl Dictionaries {
  fn best(&self, input: u32) -> Option<&DataPayload<UCharDictionaryBreakDataV1Marker>> {
    match get_language(input) {
        Language::Khmer => self.khmer,
        Language::Lao => self.lao,
        Language::Burmese => self.burmese,
        Language::Thai => self.thai,
        Language::ChineseOrJapanese => self.cj,
        _ => None,
    }

  }
}

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: experimental/segmenter/src/complex.rs [hunk 6]
```diff
@@ -13,11 +13,33 @@ use crate::DictionarySegmenter;
...
+}
+
 /// Return UTF-16 segment offset array using dictionary or lstm segmenter.
-pub fn complex_language_segment_utf16(
-    payload: &DataPayload<UCharDictionaryBreakDataV1Marker>,
-    input: &[u16],
-) -> Vec<usize> {
+///
+/// Dictionary payload has to be this order.
+///   [Khmer, Lao, Burmese, Thai]
```
Comment (robertbastian) [inline experimental/segmenter/src/complex.rs:41]: Nit: remove

File: experimental/segmenter/src/complex.rs [hunk 7]
```diff
@@ -53,10 +75,10 @@ pub fn complex_language_segment_utf16(
 }
 
 /// Return UTF-8 segment offset array using dictionary or lstm segmenter.
-pub fn complex_language_segment_str(
-    payload: &DataPayload<UCharDictionaryBreakDataV1Marker>,
-    input: &str,
-) -> Vec<usize> {
+///
+/// Dictionary payload has to be this order.
```
Comment (robertbastian) [inline experimental/segmenter/src/complex.rs:79]: Nit: remove

File: experimental/segmenter/src/line.rs [hunk 8]
```diff
@@ -99,10 +100,22 @@ impl Default for LineBreakOptions {
 pub struct LineBreakSegmenter {
     options: LineBreakOptions,
     payload: DataPayload<LineBreakDataV1Marker>,
-    dictionary_payload: DataPayload<UCharDictionaryBreakDataV1Marker>,
+    dictionary_km_payload: Option<DataPayload<UCharDictionaryBreakDataV1Marker>>,
```
Comment (robertbastian) [inline experimental/segmenter/src/line.rs:103]: Thought: would it make sense to use the `Dictionaries` structure here? You'd have to make it own its data, but you're passing it by reference anyway so there's no cloning.

File: experimental/segmenter/src/word.rs [hunk 9]
```diff
@@ -27,7 +27,11 @@ pub type WordBreakIteratorUtf16<'l, 's> = RuleBreakIterator<'l, 's, WordBreakTyp
 /// encodings. Please see the [module-level documentation](crate) for its usages.
 pub struct WordBreakSegmenter {
     payload: DataPayload<WordBreakDataV1Marker>,
-    dictionary_payload: DataPayload<UCharDictionaryBreakDataV1Marker>,
+    dictionary_cj_payload: Option<DataPayload<UCharDictionaryBreakDataV1Marker>>,
```
Comment (robertbastian) [inline experimental/segmenter/src/word.rs:30]: ditto

**Review by robertbastian — CHANGES_REQUESTED** [review]

**Review by robertbastian — APPROVED** [review]

File: provider/datagen/src/transform/segmenter/mod.rs [hunk 10]
```diff
@@ -339,9 +339,14 @@ impl SegmenterRuleProvider {
                                     properties_map[c as usize] = property_index
                                 }
                             }
+
                             continue;
                         }
 
+                        // TODO:
```
Comment (aethanyc) [inline provider/datagen/src/transform/segmenter/mod.rs:346]: Suggestion: Do we want to address this issue? If so, please file a issue and insert the issue number here. 

It seems reasonable to default use UAX29 rules for Katakana unless dictionary defines something better. (I wonder how ICU4C behaves on Katakana.)

**Review by aethanyc — DISMISSED** [review]

**Review by robertbastian — APPROVED** [review]
Please don't rebase after people have started reviewing the PR. Use merge commits instead
